### PR TITLE
Restructuring codebase

### PR DESCRIPTION
**Pull request by AlexAltea**
**DISCLAIMER: These are quite deep changes to the repository, we should probably make sure everyone agrees in the proposed changes down to the finest detail.**

Goal is simply make the platform-specific changes consistent with each other, and keeping the root-level and source folders clean of platform-specific files.

This builds correctly under Windows+VS2017 and macOS+Xcode. Testing on Windows+EWDK is pending.
There's no functionality changes, so there should be no run-time regressions.

## Changes

Generic changes:
- **Combined platform-specific folders**. Moved `{darwin,windows}/` to `platforms/{darwin,windows}/`, to avoid polluting the root folder. This is meant to reduce root-folder pollution, specially if support for more platforms (e.g. Linux) is added.
- **Combined documentation**: Moved `API.md` to `docs/api.md`, and added platform-specific usage instructions in `docs/manual-{darwin,windows}.md` to avoid polluting `README.md`, specially if support for more platforms (e.g. Linux) is added.
- **Reorganized .gitignore files**: One root-level `.gitignore` file with general rules that affect the entire source tree (e.g.: _.DS_Store_ for macOS), and platform/compiler-specific rules at `platforms/*/.gitignore` to deal with build-related files (e.g.: _*.vcxproj.user_ for Visual Studio, _build/_ for Xcode).
- **Improved consistency across platforms**: Before, each platform acted,
  - Renamed `darwin/hax_driver/com_intel_hax` to `darwin` (consistent folder depth).
  - Renamed Windows build directory from `obj` to `build` (consistent naming with macOS).

Changes on Windows:
- **Modified solution filters**: Updated solution filters/folders to match the layout of the repository {core, platforms, tests}.
- **Removed WDK 7.1 makefiles**: According to https://github.com/intel/haxm/pull/108#issuecomment-431271607, these were part of the legacy WDK 7.1 build system used in earlier HAXM releases and no longer consumed. This was replaced with WDK 10, whose Nmake2MsBuild tool generated the necessary .props files.
- **Removed WDK 10 makefiles**: There's no need of having one extra layer of indirection in our build systems: both VS and EWDK consume only _.sln_/_.vcxproj_ files which are already checked in the repository. The additions to _.vcxproj_'s are not significantly more verbose (and redundancy could still be reduced by removing duplicated statements if necessary).
- **Explicitly disabled JustMyCode**: Debug configurations enable JustMyCode which causes a [known bug](https://developercommunity.visualstudio.com/content/problem/302014/dirver-build-debugmode-checkfordebuggerjustmycode.html) in driver projects under Visual Studio 15.8.

Changes on macOS:
- **Removed readme**: Previously at `darwin\hax_driver\com_intel_hax\readme`, since its description of the codebase no longer matched the state of the codebase.

### CODE REVIEW

File: platforms/windows/haxm-core.vcxproj [hunk 1]
```diff
@@ -26,19 +26,10 @@
...
-    <IntDir Condition="'$(Platform)'=='Win32'">$(BUILD_ALT_DIR)\x86\</IntDir>
-    <OutDir>$(IntDir)</OutDir>
-  </PropertyGroup>
-  <ImportGroup Label="PreConfiguration">
-    <Import Project="$(VCTargetsPath)\WDKConversion\PreConfiguration.props" />
-  </ImportGroup>
   <PropertyGroup Label="Globals">
     <ProjectGuid>{BC80D1E0-5738-4048-A742-8A20949A6587}</ProjectGuid>
     <RootNamespace>$(MSBuildProjectName)</RootNamespace>
+    <WindowsTargetPlatformVersion>$(LatestTargetPlatformVersion)</WindowsTargetPlatformVersion>
```
[the remark this answers is not in the document]
Comment (AlexAltea) [reply platforms/windows/haxm-core.vcxproj]: > Is this autogenerated by Visual Studio?

Yes. I'm not entirely sure about the implications of that line.

File: platforms/windows/haxm-windows.vcxproj [hunk 2]
```diff
@@ -27,19 +27,10 @@
...
-    <IntDir Condition="'$(Platform)'=='Win32'">$(BUILD_ALT_DIR)\x86\</IntDir>
-    <OutDir>$(IntDir)</OutDir>
-  </PropertyGroup>
-  <ImportGroup Label="PreConfiguration">
-    <Import Project="$(VCTargetsPath)\WDKConversion\PreConfiguration.props" />
-  </ImportGroup>
   <PropertyGroup Label="Globals">
     <ProjectGuid>{BA777056-A57E-492D-A821-68D08A2DACE6}</ProjectGuid>
     <RootNamespace>$(MSBuildProjectName)</RootNamespace>
+    <WindowsTargetPlatformVersion>$(LatestTargetPlatformVersion)</WindowsTargetPlatformVersion>
```
[the remark this answers is not in the document]
Comment (AlexAltea) [reply platforms/windows/haxm-windows.vcxproj]: > Is this autogenerated by Visual Studio?

Yes. I'm not entirely sure about the implications of that line.

File: include/hax_types.h [hunk 3]
```diff
@@ -67,7 +67,7 @@
 #define HAX_PLATFORM_DARWIN
 #include "darwin/hax_types_mac.h"
 // Windows
-#elif defined(__WINNT__)
+#elif defined(WINNT)
```
[the remark this answers is not in the document]
Comment (AlexAltea) [reply include/hax_types.h]: I found it among the "inherited values" for the project's preprocessor definitions. This likely means that some internal Visual Studio .props file defines it. I just chose `WINNT` due to the similarity to the `__WINNT__` (which was by the way manually defined in *sources.props*).

I agree we should use the more common `_WIN32` macro, so I will change it.

File: platforms/windows/haxm-windows.vcxproj [hunk 4]
```diff
@@ -80,52 +71,79 @@
     <Import Project="$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props" Condition="exists('$(UserRootDir)\Microsoft.Cpp.$(Platform).user.props')" />
   </ImportGroup>
   <PropertyGroup Label="UserMacros" />
+  <PropertyGroup Label="Configuration">
+    <OutDir>$(SolutionDir)build\out\$(DDKSpec)\</OutDir>
```
[the remark this answers is not in the document]
Comment (AlexAltea) [reply platforms/windows/haxm-windows.vcxproj]: That makes much more sense, thanks for the hint! I was trying to resemble the old paths (`obj\out\Win7`), but I agree that `build\out\$(Platform)\$(ConfigurationName` is much more convenient.

**Comment by raphaelning** [comment]
Thanks, the new source tree layout looks so neat and clean!

nit: In `platforms/windows/`, there are 3 `.c` files where some of the `#include` paths are not "canonical", although that doesn't break the build:
* `hax_win.h`
* `hax_host_mem.c`
* `hax_wrapper.c`

We have started testing this PR, and have found one issue so far: the `tests` project fails the Win32 build (both Debug and Release).

```
  test_emulator.obj : error LNK2019: unresolved external symbol _em_decode_insn referenced in function "protected: void __thiscall EmulatorTest::assemble_decode(char c
onst *,unsigned __int64,unsigned int *,unsigned int *,enum em_status_t *)" (?assemble_decode@EmulatorTest@@IAEXPBD_KPAI2PAW4em_status_t@@@Z) [\path\to\haxm\platforms\wi
ndows\haxm-tests.vcxproj]
  test_emulator.obj : error LNK2019: unresolved external symbol _em_emulate_insn referenced in function "protected: void __thiscall EmulatorTest::run(char const *,stru
ct test_cpu_t const &,struct test_cpu_t const &)" (?run@EmulatorTest@@IAEXPBDABUtest_cpu_t@@1@Z) [\path\to\haxm\platforms\windows\haxm-tests.vcxproj]
  \path\to\haxm\platforms\windows\build\tests\Win32\\haxm-tests.exe : fatal error LNK1120: 2 unresolved externals [\path\to\haxm\platforms\windows\haxm-tests.vcxproj]
```

**Comment by raphaelning** [comment]
> By canonical you mean they are not relative to the current file? If so, I've just fixed it. The reason it worked is that the project (and the source.props file before), has several added include directories.

Right. I noticed that you had updated most relative `#include` paths, so I just thought you had missed a few.

> This was seemingly caused my the last PR, since the emulator tests include emulate.h, which relies on HAX_ARCH* macros, which are included from include/hax_types.h if HAX_TESTS is not defined, which it is.

I see. I thought we had tested the Win32 build before merging #110, but perhaps we overlooked the `tests` build failure.

> In the future we probably should make a clear cut between kernel and userspace: maybe by defining macros HAX_KERNEL and HAX_USER (instead of HAX_TESTS) that lets us separate code better.

I agree. Maybe this can help us solve another problem (#9): that `hax_interface.h` is actually not shared between HAXM (kernel) and QEMU (user space). There are a few issues to fix, but making `hax_types.h` portable would be a key step.

**Comment by AlexAltea** [comment]
>  Maybe this can help us solve another problem (#9): that hax_interface.h is actually not shared between HAXM (kernel) and QEMU (user space).

Yes, my thoughts exactly. :-)

**Pull request merged**